subcomponents/assemblies are not supported in `mapdl.components`

As the title.

This test fails.

```py
def test_subcomponents(mapdl, cleared):
    mapdl.prep7()

    with mapdl.non_interactive:
        mapdl.run("*do,i,1,10")
        mapdl.run("nsel,none")
        mapdl.run("n,i,i,0,0")
        mapdl.cm(f"NODE_%i%", "NODE")
        mapdl.run("*enddo")

    mapdl.allsel()

    for i in range(1, 11):
        assert f"NODE_{i}" in str(mapdl.components)
        assert len(mapdl.components[f"NODE_{i}"]) == 1
        assert mapdl.components[f"NODE_{i}"][0] == i
        
    
    mapdl.cmgrp("assembly", "NODE_1", "NODE_2", "NODE_3", "NODE_4", "NODE_5", "NODE_6", "NODE_7", "NODE_8", "NODE_9", "NODE_10")

    assert "ASSEMBLY" in str(mapdl.components)
    assert len(mapdl.components["assembly"]) == 10
```

and fails with:

```pytb

>       assert "ASSEMBLY" in str(mapdl.components)
                             ^^^^^^^^^^^^^^^^^^^^^

tests/test_component.py:454: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/ansys/mapdl/core/component.py:469: in __repr__
    if self._comp:
       ^^^^^^^^^^
src/ansys/mapdl/core/component.py:327: in _comp
    self.__comp = self._mapdl._parse_cmlist()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
src/ansys/mapdl/core/mapdl_core.py:3202: in _parse_cmlist
    return _parse_cmlist(cmlist)
           ^^^^^^^^^^^^^^^^^^^^^
src/ansys/mapdl/core/component.py:649: in _parse_cmlist
    extract(each_line, 0): extract_line(each_line)
                           ^^^^^^^^^^^^^^^^^^^^^^^
src/ansys/mapdl/core/component.py:646: in extract_line
    return extract(each_line, 1)
           ^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

each_line = '                                          NODE_2                          ', ind = 1

    def extract(each_line, ind):
>       return each_line.split()[ind].strip()
               ^^^^^^^^^^^^^^^^^^^^^^
E       IndexError: list index out of range

src/ansys/mapdl/core/component.py:640: IndexError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
```

because this is the `cmlist` output:

```LIST ALL SELECTED COMPONENTS
   *****MAPDL VERIFICATION RUN ONLY*****
     DO NOT USE RESULTS FOR PRODUCTION

    NAME                            TYPE      SUBCOMPONENTS

  ASSEMBLY                          LVL1  NODE_1                          
                                          NODE_2                          
                                          NODE_3                          
                                          NODE_4                          
                                          NODE_5                          
                                          NODE_6                          
                                          NODE_7                          
                                          NODE_8                          
  NODE_1                            NODE
  NODE_10                           NODE
  NODE_2                            NODE
  NODE_3                            NODE
  NODE_4                            NODE
  NODE_5                            NODE
  NODE_6                            NODE
  NODE_7                            NODE
  NODE_8                            NODE
  NODE_9                            NODE
```

the problem is the lines below the `type` definition (`lvl1`), aka `NODE_2`, `NODE_3`, etc

--- a/doc/changelog.d/4647.fixed.md
+++ b/doc/changelog.d/4647.fixed.md
@@ -1,0 +1,1 @@
+Enhance component parsing to support assembly-level types and subcomponent extraction

--- a/src/ansys/mapdl/core/commands.py
+++ b/src/ansys/mapdl/core/commands.py
@@ -989,7 +989,7 @@ class ComponentListing(CommandListingOutput):
         from ansys.mapdl.core.component import _parse_cmlist  # type: ignore
 
         # To keep same API as commands
-        return np.array(list(_parse_cmlist(self).keys()))
+        return np.array(list(_parse_cmlist(self)[0].keys()))
 
 
 class StringWithLiteralRepr(str):

--- a/src/ansys/mapdl/core/component.py
+++ b/src/ansys/mapdl/core/component.py
@@ -279,6 +279,7 @@ class ComponentManager:
 
         self._mapdl_weakref: weakref.ReferenceType[Mapdl] = weakref.ref(mapdl)
         self.__comp: UNDERLYING_DICT = {}
+        self._subcomponents: Dict[str, List[str]] = {}
         self._update_always: bool = True
         self._autoselect_components: bool = (
             False  # if True, PyMAPDL will try to select the CM first if it does not appear in the CMLIST output.
@@ -327,7 +328,7 @@ class ComponentManager:
     def _comp(self) -> UNDERLYING_DICT:
         """Dictionary with components names and types."""
         if self.__comp is None or self._update_always:
-            self.__comp = self._mapdl._parse_cmlist()
+            self.__comp, self._subcomponents = self._mapdl._parse_cmlist()
         return self.__comp
 
     @_comp.setter
@@ -365,9 +366,9 @@ class ComponentManager:
                 )
 
         if re.match(r"LVL\d+$", cmtype, re.IGNORECASE):
-            # Assembly: return subcomponent names with the assembly-level type.
-            cmlist_str = self._mapdl.cmlist(key, 1)
-            subcomp_names = _parse_assembly_subcomponents(cmlist_str, key)
+            # Assembly: subcomponent names were already collected during the
+            # initial _parse_cmlist pass — no extra MAPDL call needed.
+            subcomp_names = self._subcomponents.get(key.upper(), [])
             if forced_to_select:
                 self._mapdl.cmsel("U", key)
             return Component(cmtype, subcomp_names)  # type: ignore[arg-type]
@@ -620,50 +621,22 @@ def _parse_cmlist_indiv(
     return items_int
 
 
-def _parse_assembly_subcomponents(cmlist_output: str, assembly_name: str) -> List[str]:
-    """Extract subcomponent names from a cmlist output block for an assembly.
-
-    Parameters
-    ----------
-    cmlist_output : str
-        Raw output from ``CMLIST`` for the assembly.
-    assembly_name : str
-        Name of the assembly (case-insensitive).
+def _parse_cmlist(
+    cmlist: Optional[str] = None,
+) -> Tuple[Dict[str, Any], Dict[str, List[str]]]:
+    """Parse the output of the CMLIST command.
 
     Returns
     -------
-    list of str
-        Names of the subcomponents belonging to the assembly.
+    tuple
+        A pair ``(comp_dict, assembly_subcomps)`` where *comp_dict* maps
+        component names to their type (or ``[selection, type]`` when the
+        selection column is present) and *assembly_subcomps* maps each
+        assembly name to the list of its subcomponent names.
     """
-    subcomps: List[str] = []
-    in_assembly = False
-    for line in cmlist_output.splitlines():
-        parts = line.split()
-        if not parts:
-            if in_assembly:
-                break
-            continue
-        if not in_assembly:
-            if (
-                parts[0].upper() == assembly_name.upper()
-                and len(parts) >= 2
-                and re.match(r"LVL\d+", parts[1], re.IGNORECASE)
-            ):
-                in_assembly = True
-                if len(parts) >= 3:
-                    subcomps.append(parts[2])
-        else:
-            if len(parts) == 1:
-                subcomps.append(parts[0])
-            else:
-                break  # Next component entry starts
-    return subcomps
-
-
-def _parse_cmlist(cmlist: Optional[str] = None) -> Dict[str, Any]:
     include_selection = False
     if cmlist is None:
-        return {}
+        return {}, {}
     if re.search(r"NAME\s+TYPE\s+SUBCOMPONENTS", cmlist):
         # header
         #  "NAME                            TYPE      SUBCOMPONENTS"
@@ -685,19 +658,35 @@ def _parse_cmlist(cmlist: Optional[str] = None) -> Dict[str, Any]:
     else:
         raise ValueError("The format of the CMLIST output is not recognaised.")
 
-    cmlist = "\n".join(blocks)
+    # Index of the TYPE token (0-based within split tokens of a component line)
+    type_idx = 2 if include_selection else 1
 
-    def extract(each_line, ind):
-        return each_line.split()[ind].strip()
+    comp_dict: Dict[str, Any] = {}
+    assembly_subcomps: Dict[str, List[str]] = {}
+    current_assembly: Optional[str] = None
 
-    def extract_line(each_line):
-        if include_selection:
-            return [extract(each_line, 1), extract(each_line, 2)]
-        else:
-            return extract(each_line, 1)
+    for line in "\n".join(blocks).splitlines():
+        parts = line.split()
+        if not parts:
+            current_assembly = None
+            continue
+        if len(parts) >= 2:
+            # Regular component line or assembly header line
+            name = parts[0]
+            value: Any = [parts[1], parts[2]] if include_selection else parts[1]
+            comp_dict[name] = value
+            if re.match(r"LVL\d+$", parts[type_idx], re.IGNORECASE):
+                current_assembly = name.upper()
+                assembly_subcomps[current_assembly] = []
+                # First subcomponent may appear on the same line
+                first_subcomp_idx = type_idx + 1
+                if len(parts) > first_subcomp_idx:
+                    assembly_subcomps[current_assembly].append(parts[first_subcomp_idx])
+            else:
+                current_assembly = None
+        elif current_assembly is not None:
+            # Continuation line: a single subcomponent name belonging to the
+            # current assembly block
+            assembly_subcomps[current_assembly].append(parts[0])
 
-    return {
-        extract(each_line, 0): extract_line(each_line)
-        for each_line in cmlist.splitlines()
-        if each_line and len(each_line.split()) >= 2
-    }
+    return comp_dict, assembly_subcomps

--- a/src/ansys/mapdl/core/component.py
+++ b/src/ansys/mapdl/core/component.py
@@ -164,8 +164,13 @@ class Component(tuple):
         elif items_ is None:
             items_ = tuple()
         elif isinstance(items_, (int, str)):
-            # Handle single values
-            items_ = (int(items_),) if isinstance(items_, int) else (items_,)
+            # Handle single values.
+            # For assembly types (LVL*) the single item is a component name
+            # string; for all entity types numeric strings must be cast to int.
+            if isinstance(items_, str) and re.match(r"LVL\d+$", type_, re.IGNORECASE):
+                items_ = (items_,)
+            else:
+                items_ = (int(items_),)
         elif not isinstance(items_, tuple):
             # Fallback for any iterable
             try:

--- a/src/ansys/mapdl/core/component.py
+++ b/src/ansys/mapdl/core/component.py
@@ -149,7 +149,10 @@ class Component(tuple):
         items_: Union[Tuple[int, ...], List[int], "Component", NDArray, int, str, None],
     ) -> "Component":
         """Create a new Component instance."""
-        if not isinstance(type_, str) or type_.upper() not in VALID_ENTITIES:
+        if not isinstance(type_, str) or (
+            type_.upper() not in VALID_ENTITIES
+            and not re.match(r"LVL\d+$", type_, re.IGNORECASE)
+        ):
             raise ValueError(
                 f"The value '{type_}' is not allowed for 'type' definition."
             )
@@ -162,7 +165,7 @@ class Component(tuple):
             items_ = tuple()
         elif isinstance(items_, (int, str)):
             # Handle single values
-            items_ = (int(items_),) if isinstance(items_, (int, str)) else tuple()
+            items_ = (int(items_),) if isinstance(items_, int) else (items_,)
         elif not isinstance(items_, tuple):
             # Fallback for any iterable
             try:
@@ -360,6 +363,14 @@ class ComponentManager:
                 raise ComponentDoesNotExits(
                     f"The component named '{key}' does not exist in the MAPDL instance."
                 )
+
+        if re.match(r"LVL\d+$", cmtype, re.IGNORECASE):
+            # Assembly: return subcomponent names with the assembly-level type.
+            cmlist_str = self._mapdl.cmlist(key, 1)
+            subcomp_names = _parse_assembly_subcomponents(cmlist_str, key)
+            if forced_to_select:
+                self._mapdl.cmsel("U", key)
+            return Component(cmtype, subcomp_names)  # type: ignore[arg-type]
 
         output: List[int] = self._mapdl._parse_cmlist_indiv(key, cmtype)
         if forced_to_select:
@@ -609,6 +620,46 @@ def _parse_cmlist_indiv(
     return items_int
 
 
+def _parse_assembly_subcomponents(cmlist_output: str, assembly_name: str) -> List[str]:
+    """Extract subcomponent names from a cmlist output block for an assembly.
+
+    Parameters
+    ----------
+    cmlist_output : str
+        Raw output from ``CMLIST`` for the assembly.
+    assembly_name : str
+        Name of the assembly (case-insensitive).
+
+    Returns
+    -------
+    list of str
+        Names of the subcomponents belonging to the assembly.
+    """
+    subcomps: List[str] = []
+    in_assembly = False
+    for line in cmlist_output.splitlines():
+        parts = line.split()
+        if not parts:
+            if in_assembly:
+                break
+            continue
+        if not in_assembly:
+            if (
+                parts[0].upper() == assembly_name.upper()
+                and len(parts) >= 2
+                and re.match(r"LVL\d+", parts[1], re.IGNORECASE)
+            ):
+                in_assembly = True
+                if len(parts) >= 3:
+                    subcomps.append(parts[2])
+        else:
+            if len(parts) == 1:
+                subcomps.append(parts[0])
+            else:
+                break  # Next component entry starts
+    return subcomps
+
+
 def _parse_cmlist(cmlist: Optional[str] = None) -> Dict[str, Any]:
     include_selection = False
     if cmlist is None:
@@ -648,5 +699,5 @@ def _parse_cmlist(cmlist: Optional[str] = None) -> Dict[str, Any]:
     return {
         extract(each_line, 0): extract_line(each_line)
         for each_line in cmlist.splitlines()
-        if each_line
+        if each_line and len(each_line.split()) >= 2
     }

--- a/src/ansys/mapdl/core/component.py
+++ b/src/ansys/mapdl/core/component.py
@@ -658,7 +658,7 @@ def _parse_cmlist(
         )
         # Using `cmlist,all` which also includes selection
         include_selection = True
-    elif "LIST ALL SELECTED COMPONENTS":
+    elif "LIST ALL SELECTED COMPONENTS" in cmlist:
         blocks = cmlist.splitlines()[1:]
     else:
         raise ValueError("The format of the CMLIST output is not recognaised.")

--- a/src/ansys/mapdl/core/component.py
+++ b/src/ansys/mapdl/core/component.py
@@ -661,7 +661,7 @@ def _parse_cmlist(
     elif "LIST ALL SELECTED COMPONENTS" in cmlist:
         blocks = cmlist.splitlines()[1:]
     else:
-        raise ValueError("The format of the CMLIST output is not recognaised.")
+        raise ValueError("The format of the CMLIST output is not recognized.")
 
     # Index of the TYPE token (0-based within split tokens of a component line)
     type_idx = 2 if include_selection else 1

--- a/src/ansys/mapdl/core/mapdl_core.py
+++ b/src/ansys/mapdl/core/mapdl_core.py
@@ -3193,7 +3193,9 @@ class _MapdlCore(Commands):
                 self._parent()._run("/nopr")
             self._parent()._mute = self._previous_mute
 
-    def _parse_cmlist(self, cmlist: Optional[str] = None) -> Dict[str, Any]:
+    def _parse_cmlist(
+        self, cmlist: Optional[str] = None
+    ) -> Tuple[Dict[str, Any], Dict[str, List[str]]]:
         from ansys.mapdl.core.component import _parse_cmlist
 
         if not cmlist:

--- a/tests/test_component.py
+++ b/tests/test_component.py
@@ -81,6 +81,35 @@ def test_component_wrong_init():
         Component("asdf", [1, 2, 3])
 
 
+def test_component_single_int_item():
+    cm = Component("NODE", 5)
+    assert cm == (5,)
+    assert isinstance(cm[0], int)
+
+
+def test_component_single_numeric_str_item():
+    cm = Component("NODE", "5")
+    assert cm == (5,)
+    assert isinstance(cm[0], int)
+
+
+def test_component_single_str_item_lvl():
+    cm = Component("LVL1", "NODE_1")
+    assert cm == ("NODE_1",)
+    assert isinstance(cm[0], str)
+    assert cm.type == "LVL1"
+
+
+def test_component_generic_iterable_item():
+    cm = Component("NODE", (x for x in [1, 2, 3]))
+    assert cm == (1, 2, 3)
+
+
+def test_component_non_iterable_item_fallback():
+    cm = Component("NODE", object())
+    assert cm == ()
+
+
 def test_logger(mapdl, cleared):
     assert mapdl.components.logger == mapdl.logger
 

--- a/tests/test_component.py
+++ b/tests/test_component.py
@@ -444,7 +444,7 @@ def test_subcomponents(mapdl, cleared):
         mapdl.run("*do,i,1,10")
         mapdl.run("nsel,none")
         mapdl.run("n,i,i,0,0")
-        mapdl.cm(f"NODE_%i%", "NODE")
+        mapdl.cm("NODE_%i%", "NODE")
         mapdl.run("*enddo")
 
     mapdl.allsel()

--- a/tests/test_component.py
+++ b/tests/test_component.py
@@ -418,8 +418,9 @@ def test_big_component(mapdl, cleared):
 def test_lots_of_components(mapdl, cleared):
     mapdl.prep7()
 
+    n_cmp = 999
     with mapdl.non_interactive:
-        mapdl.run("*do,i,1,999")
+        mapdl.run(f"*do,i,1,{n_cmp}")
         mapdl.run("nsel,none")
         mapdl.run("n,i,i,0,0")
         mapdl.run(
@@ -427,11 +428,11 @@ def test_lots_of_components(mapdl, cleared):
         )
         mapdl.run("*enddo")
 
-    assert len(mapdl.components) == 999
+    assert len(mapdl.components) == n_cmp
     cmp = str(mapdl.components)
-    for i in range(1, 1000):
+    for i in range(1, n_cmp + 1):
         assert (
-            f"NODE_{i}_U760_P79420_SAPH400_HINGE_BODY_SIDE_UPR_FR_DR_LH_123_Aasdf_asdfa_123_"
+            f"NODE_{i}_U760_P79420_SAPH400_HINGE_BODY_SIDE_UPR_FR_DR_LH_123_Aasdf_asdfa_123_".upper()
             in cmp
         )
 

--- a/tests/test_component.py
+++ b/tests/test_component.py
@@ -413,3 +413,69 @@ def test_big_component(mapdl, cleared):
     assert "MANY_NODES" in str(mapdl.components)
     assert len(mapdl.components["many_nodes"]) == 999
     assert all(isinstance(item, int) for item in mapdl.components["many_nodes"])
+
+
+def test_lots_of_components(mapdl, cleared):
+    mapdl.prep7()
+
+    with mapdl.non_interactive:
+        mapdl.run("*do,i,1,999")
+        mapdl.run("nsel,none")
+        mapdl.run("n,i,i,0,0")
+        mapdl.run(
+            "cm,NODE_%i%_U760_P79420_SAPH400_HINGE_BODY_SIDE_UPR_FR_DR_LH_123_Aasdf_asdfa_123_, node"
+        )
+        mapdl.run("*enddo")
+
+    assert len(mapdl.components) == 999
+    cmp = mapdl.components
+    for i in range(1, 1000):
+        assert (
+            f"NODE_{i}_U760_P79420_SAPH400_HINGE_BODY_SIDE_UPR_FR_DR_LH_123_Aasdf_asdfa_123_"
+            in cmp
+        )
+
+
+def test_subcomponents(mapdl, cleared):
+    mapdl.prep7()
+
+    with mapdl.non_interactive:
+        mapdl.run("*do,i,1,10")
+        mapdl.run("nsel,none")
+        mapdl.run("n,i,i,0,0")
+        mapdl.cm(f"NODE_%i%", "NODE")
+        mapdl.run("*enddo")
+
+    mapdl.allsel()
+
+    for i in range(1, 11):
+        assert f"NODE_{i}" in str(mapdl.components)
+        assert len(mapdl.components[f"NODE_{i}"]) == 1
+        assert mapdl.components[f"NODE_{i}"][0] == i
+
+    mapdl.cmgrp(
+        "assembly",
+        "NODE_1",
+        "NODE_2",
+        "NODE_3",
+        "NODE_4",
+        "NODE_5",
+        "NODE_6",
+        "NODE_7",
+        "NODE_8",
+    )
+
+    print(mapdl.components)  # This will trigger COMP parsing
+
+    assert "ASSEMBLY" in str(mapdl.components)
+    assert len(mapdl.components["assembly"]) == 8
+    assert "LVL" in mapdl.components["assembly"].type
+    assert mapdl.components["assembly"].type == "LVL1"
+    assert "NODE_1" in mapdl.components["assembly"].items
+    assert "NODE_2" in mapdl.components["assembly"].items
+    assert "NODE_3" in mapdl.components["assembly"].items
+    assert "NODE_4" in mapdl.components["assembly"].items
+    assert "NODE_5" in mapdl.components["assembly"].items
+    assert "NODE_6" in mapdl.components["assembly"].items
+    assert "NODE_7" in mapdl.components["assembly"].items
+    assert "NODE_8" in mapdl.components["assembly"].items

--- a/tests/test_component.py
+++ b/tests/test_component.py
@@ -428,7 +428,7 @@ def test_lots_of_components(mapdl, cleared):
         mapdl.run("*enddo")
 
     assert len(mapdl.components) == 999
-    cmp = mapdl.components
+    cmp = str(mapdl.components)
     for i in range(1, 1000):
         assert (
             f"NODE_{i}_U760_P79420_SAPH400_HINGE_BODY_SIDE_UPR_FR_DR_LH_123_Aasdf_asdfa_123_"

--- a/tests/test_component.py
+++ b/tests/test_component.py
@@ -478,8 +478,9 @@ def test_subcomponents(mapdl, cleared):
 
     mapdl.allsel()
 
+    cmp_repr = str(mapdl.components)
     for i in range(1, 11):
-        assert f"NODE_{i}" in str(mapdl.components)
+        assert f"NODE_{i}" in cmp_repr
         assert len(mapdl.components[f"NODE_{i}"]) == 1
         assert mapdl.components[f"NODE_{i}"][0] == i
 
@@ -495,7 +496,7 @@ def test_subcomponents(mapdl, cleared):
         "NODE_8",
     )
 
-    print(mapdl.components)  # This will trigger COMP parsing
+    _ = str(mapdl.components)  # Trigger component parsing
 
     assert "ASSEMBLY" in str(mapdl.components)
     assert len(mapdl.components["assembly"]) == 8

